File: 49-group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // P: group anagrams together given an array of strings
        // return some sort of array of arrays with strings grouped together 
        // based on if they share the same set of letters
        // -- lower case English letters
        // E: provided by problem
        // D: HashMap of ArrayLists
        // -- key: word with letters in sorted order
        // -- value: ArrayList that contains words that have the same
        // -- exact letters
        // A: a general loop, loop throguh array of strings
        // sort the word, without changing the original word ofc
        // if sorted word exists:
        // add word to the arrayList that it maps to
        // else:
        // create array list, and add
        // og word to arraylist
        // put that new sorted word
        // to hashmap,  with arraylist as value
        // then at the end, return an array of size, number of keys, 
        // and each ArrayList as an object

        HashMap<String, ArrayList<String>> 
            sortedListToWordsArray = new HashMap<>();
        for(int i = 0; i < strs.length; i++){
            char[] charArray = strs[i].toCharArray();
            Arrays.sort(charArray);
            String sortedString = new String(charArray);
            if(sortedListToWordsArray.containsKey(sortedString)){
                sortedListToWordsArray.get(sortedString).add(strs[i]);
            }
            else {
                ArrayList<String> wordsArray = new ArrayList<>();
                wordsArray.add(strs[i]);
                sortedListToWordsArray.put(sortedString, wordsArray);
            }
        }
        List<List<String>> groupedAnagrams = new ArrayList<>();
        Object[] keyArray = sortedListToWordsArray.keySet().toArray();
        for(int i = 0; i < keyArray.length; i++){
            groupedAnagrams.add(
                sortedListToWordsArray.get(keyArray[i])
            );
        }
        return groupedAnagrams;
    }
}

File: 49-group-anagrams/group-anagrams.py
class Solution:
    """
    P:
        - given a List of str called strs
        - want: a List of List of str vals - result
            - want to group the anagrams together, meaning that the words
            in a result[i] are anagrams with each other (we can rearrange them to get any other
            word in the group)
            - populate ofc from strs 
        - constraints:
            - the List - strs is non-zero length up to 10^4
            - any individual str element can be 0 length (empty string) but no longer than 100 chars
            - str[i] consists of lowercase letters
    E:
        - the examples make sense overall
    D:
        - Could use a map, which would map from word to some representation
            - either a representation of letter counts
            - or - an ordered list of the characters themselves sorted
        - OR: a map that takes a representation and as a value
        collects the words that share the same representation
    A:
        - we know for a fact that anagrams will have the same exact letter counts
        - IDEA 1:
            - init reps_to_word = {}
            - for word in strs:
                - word_rep = get_representation(word)
                - if word_rep in reps_to_word:
                    - reps_to_word[word_rep].append(word)

            - def get_representation(word):
                - counts = {}
                - for char in word:
                    - counts[char] = counts.get(char, 0) + 1
                    

    C:
    """
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        result = [] # elements stored are List[str] where all elements of that interior list are anagrams
        reps_to_word = {} # maps from a word_rep to a list of the words that contain that rep
        def get_representation(word): 
            """
            Gets a representation of the word, 
            mapping from each char to the word with
            letters sorted in order
            """
            # word_rep = {}
            # for char in word:
            #     word_rep[char] = word_rep.get(char, 0) + 1
            # rep = []
            # for char in word_rep.keys():
            #     rep.append((char, word_rep[char]))
            # rep = tuple(sorted(rep, key=lambda char_count: char_count[0]))
            rep = ''.join(sorted(word))
            return rep
        
        # populate word counts representations
        for word in strs:
            word_rep = get_representation(word)
            if word_rep in reps_to_word:
                reps_to_word[word_rep].append(word)
            else:
                reps_to_word[word_rep] = [word]
        for key in reps_to_word.keys():
            result.append(reps_to_word[key])
        return result